File: 1차/2주차/정병현/광고 삽입.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;

string solution(string play_time, string adv_time, vector<string> logs) {
    int n = logs.size();
    char c;
    int h, m, s, ptime, atime;
    stringstream ssin; // use sstream for easy parsing
    vector<int> times, watching;
    string ans = "";
    ///// get playtime
    ssin.str(play_time);
    ssin >> h >> c >> m >> c >> s;
    ssin.clear();
    ptime = 3600*h + 60*m + s;
    
    times.resize(ptime+1);
    watching.resize(ptime+1);
    ///// get advertising time
    ssin.str(adv_time);
    ssin >> h >> c >> m >> c >> s;
    ssin.clear();
    atime = 3600*h + 60*m + s;
    // cout << "ptime : " << ptime << "\n";
    // cout << "atime : " << atime << "\n";
    for(int i = 0; i < n; i++){ // make differentiation array of watching people
        ssin.str(logs[i]);
        ssin >> h >> c >> m >> c >> s >> c;
        times[3600*h + 60*m + s]++;
        ssin >> h >> c >> m >> c >> s;
        ssin.clear();
        times[3600*h + 60*m + s]--;
    }
    watching[0] = times[0];
    //making sumarray
    for(int i = 1; i <= ptime; i++)
        watching[i] = watching[i-1] + times[i];
    // sliding window
    ll max_ad = 0, max_time = 0, now_ad = 0;
    for(int i = 0; i < atime; i++)
        now_ad += watching[i];
    max_ad = now_ad;
    for(int i = 0; i <= ptime - atime; i++){
        now_ad -= watching[i];
        now_ad += watching[i+atime];
        if(now_ad > max_ad){
            max_ad = now_ad;
            max_time = i+1;
            //cout << i << "\n";
        }
    }
    h = max_time / 3600;
    m = (max_time / 60) - 60*h;
    s = max_time % 60;
    if(h < 10)
        ans += "0";
    ans += to_string(h)+":";
    if(m < 10)
        ans += "0";
    ans += to_string(m)+":";
    if(s < 10)
        ans += "0";
    ans += to_string(s);
    return ans;
}

File: 1차/2주차/이수민/광고 삽입.py
def solution(play_time, adv_time, logs):
    play_time = str_to_int(play_time)        # 1
    adv_time = str_to_int(adv_time)               
    all_time = [0 for i in range(play_time + 1)]

    for l in logs:                           # 2
        start, end = l.split('-')
        start = str_to_int(start)
        end = str_to_int(end)
        all_time[start] += 1
        all_time[end] -= 1

    for i in range(1, len(all_time)):       # 3
        all_time[i] = all_time[i] + all_time[i - 1]

    for i in range(1, len(all_time)):       # 4
        all_time[i] = all_time[i] + all_time[i - 1]

    most_view = 0                           # 5
    max_time = 0                          
    for i in range(adv_time - 1, play_time):
        if i >= adv_time:
            if most_view < all_time[i] - all_time[i - adv_time]:
                most_view = all_time[i] - all_time[i - adv_time]
                max_time = i - adv_time + 1
        else:
            if most_view < all_time[i]:
                most_view = all_time[i]
                max_time = i - adv_time + 1

    return int_to_str(max_time)


def str_to_int(time):
    h, m, s = time.split(':')
    return int(h) * 3600 + int(m) * 60 + int(s)


def int_to_str(time):
    h = time // 3600
    h = '0' + str(h) if h < 10 else str(h)
    time = time % 3600
    m = time // 60
    m = '0' + str(m) if m < 10 else str(m)
    time = time % 60
    s = '0' + str(time) if time < 10 else str(time)
    return h + ':' + m + ':' + s



####all 구현으로 풀다가 시간제한이 걸리는 것 보고
####아.. 이건 구현이 아니라 어느 알고리즘을 써야하는 거구나 깨달아서
####그때 바로 답을 본 문제..
####DP...어렵다.. 답을 이해하는데도 어느정도 시간이 걸렸었다..